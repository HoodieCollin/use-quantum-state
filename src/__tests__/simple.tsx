import { expect, it } from '@jest/globals';
import '@testing-library/jest-dom';
import { act, render, screen } from '@testing-library/react';

import { BaseHTMLAttributes, useEffect, useRef } from 'react';
import { createQuantumContext } from '..';

function createTestMaterials() {
  const [TestProvider, useQuantumState] = createQuantumContext({ a: 1, b: 2 });

  return { TestProvider, TestComponent, useQuantumState };

  function TestComponent(props: BaseHTMLAttributes<HTMLDivElement>) {
    const [value, setValue] = useQuantumState('a');

    return (
      <div
        {...props}
        data-value={value}
        onClick={() => {
          setValue((prev) => prev + 1);
        }}
      />
    );
  }
}

it('renders without crashing', () => {
  const { TestProvider, TestComponent } = createTestMaterials();

  const { asFragment } = render(
    <TestProvider>
      <TestComponent />
    </TestProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('supports custom values', () => {
  const { TestProvider, TestComponent } = createTestMaterials();

  const { asFragment } = render(
    <TestProvider value={{ a: 3, b: 4 }}>
      <TestComponent />
    </TestProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('supports synced peers', () => {
  const { TestProvider, TestComponent, useQuantumState } =
    createTestMaterials();

  const { asFragment } = render(
    <TestProvider>
      <TestComponent>Item</TestComponent>
      <OtherComponent />
    </TestProvider>
  );

  expect(asFragment()).toMatchSnapshot();

  act(() => {
    screen.getByText('Item').click();
  });

  expect(asFragment()).toMatchSnapshot();

  function OtherComponent() {
    const [value] = useQuantumState('a');

    return <div>{value}</div>;
  }
});

it('does not rerender unrelated peers', () => {
  const { TestProvider, TestComponent, useQuantumState } =
    createTestMaterials();

  const { asFragment } = render(
    <TestProvider>
      <TestComponent>Item</TestComponent>
      <OtherComponent />
    </TestProvider>
  );

  expect(asFragment()).toMatchSnapshot();

  act(() => {
    screen.getByText('Item').click();
  });

  expect(asFragment()).toMatchSnapshot();

  function OtherComponent() {
    const [_] = useQuantumState('b');
    const nRef = useRef(0);

    useEffect(() => {
      nRef.current++;
    });

    return <div>Unrelated | count: {nRef.current}</div>;
  }
});
