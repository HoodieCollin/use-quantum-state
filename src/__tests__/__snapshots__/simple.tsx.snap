// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not rerender unrelated peers 1`] = `
<DocumentFragment>
  <div
    data-value="1"
  >
    Item
  </div>
  <div>
    Unrelated | count: 0
  </div>
</DocumentFragment>
`;

exports[`does not rerender unrelated peers 2`] = `
<DocumentFragment>
  <div
    data-value="2"
  >
    Item
  </div>
  <div>
    Unrelated | count: 0
  </div>
</DocumentFragment>
`;

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    data-value="1"
  />
</DocumentFragment>
`;

exports[`supports custom values 1`] = `
<DocumentFragment>
  <div
    data-value="3"
  />
</DocumentFragment>
`;

exports[`supports synced peers 1`] = `
<DocumentFragment>
  <div
    data-value="1"
  >
    Item
  </div>
  <div>
    1
  </div>
</DocumentFragment>
`;

exports[`supports synced peers 2`] = `
<DocumentFragment>
  <div
    data-value="2"
  >
    Item
  </div>
  <div>
    2
  </div>
</DocumentFragment>
`;
